packages/plugin-ext/src/main/browser/terminal-main.ts

```
@@ -0,0 +1,105 @@
+import { TerminalService } from "@theia/terminal/lib/browser/base/terminal-service";
+import { TerminalWidget, TerminalWidgetOptions } from "@theia/terminal/lib/browser/base/terminal-widget";
+import { RPCProtocol } from "../../api/rpc-protocol";
+import { ApplicationShell } from "@theia/core/lib/browser";
+
+/**
+ * Plugin api service allows working with terminal emulator.
+ */
+export class TerminalServiceMainImpl implements TerminalServiceMain {
+
+    private readonly terminalService: TerminalService;
+    private readonly shell: ApplicationShell;
+    protected readonly terminals = new Map<number, TerminalWidget>();
+    private readonly extProxy: TerminalServiceExt;
+    private terminalNumber = 0;
+    private readonly TERM_ID_PREFIX = "plugin-terminal-";
+
+    constructor(container: interfaces.Container, rpc: RPCProtocol) {
+        this.terminalService = container.get(TerminalService);
+        this.shell = container.get(ApplicationShell);
+        this.extProxy = rpc.getProxy(MAIN_RPC_CONTEXT.TERMINAL_EXT);
+    }
+
+    async $createTerminal(options: TerminalOptions): Promise<number> {
+        const counter = this.terminalNumber++;
+        const termWidgetOptions: TerminalWidgetOptions = {
+            title: options.name,
+            shellPath: options.shellPath,
+            shellArgs: options.shellArgs,
+            cwd: options.cwd,
+            env: options.env,
+            destroyTermOnClose: true,
+            useServerTitle: false,
+            id: this.TERM_ID_PREFIX + counter
+        };
+        let id: number;
+        try {
+            const termWidget = await this.terminalService.newTerminal(termWidgetOptions);
+            id = await termWidget.start();
+            this.terminals.set(id, termWidget);
+            termWidget.onTerminalDidClose(() => {
+                this.extProxy.$terminalClosed(id);
+            });
+        } catch (error) {
+            return Promise.reject("Failed to create terminal " + error);
```

should we throw an Error ?